### PR TITLE
docs: add Differential Loading section to deployment guide

### PR DESCRIPTION
**Pull request by brandonroberts**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by santoshyadav198613** [comment]
Hi @brandonroberts ,

There is a small type opened https://github.com/angular/angular/issues/29935
you can close with this.

File: aio/content/guide/deployment.md [hunk 1]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
 
 When the `base` tag is mis-configured, the app fails to load and the browser console displays `404 - Not Found` errors
 for the missing files. Look at where it _tried_ to find those files and adjust the base tag appropriately.
+
+## Differential Loading
+
+When building web applications, making sure your application is compatible with the majority of browsers is a goal. Even as JavaScript continues to evolve, with new features being introduced, all browsers are not updated with support for these new features at the same pace. This is where compilation and [polyfills](guide/browser-support#polyfills) come in. The code you write in development using TypeScript is compiled and bundled into a format that is compatible with most browsers, commonly known as ES5. Polyfills are used bridge the gap, providing functionality that simply doesn't exist in some legacy browsers.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: all browsers are not updated -> not all browsers are updated

File: aio/content/guide/deployment.md [hunk 2]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
 for the missing files. Look at where it _tried_ to find those files and adjust the base tag appropriately.
+
+## Differential Loading
+
+When building web applications, making sure your application is compatible with the majority of browsers is a goal. Even as JavaScript continues to evolve, with new features being introduced, all browsers are not updated with support for these new features at the same pace. This is where compilation and [polyfills](guide/browser-support#polyfills) come in. The code you write in development using TypeScript is compiled and bundled into a format that is compatible with most browsers, commonly known as ES5. Polyfills are used bridge the gap, providing functionality that simply doesn't exist in some legacy browsers. 
+
+There is a cost to ensure this browser compatibility, and it comes in the form of bundle size. All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't. To maximize compatibility, you ship a single bundle that includes all your transpiled code, plus any polyfills that may be needed.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: form of *larger* bundle size?

File: aio/content/guide/deployment.md [hunk 3]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+There is a cost to ensure this browser compatibility, and it comes in the form of bundle size. All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't. To maximize compatibility, you ship a single bundle that includes all your transpiled code, plus any polyfills that may be needed. 
+
+Your web application shouldn't have to keep paying the penalty of increased bundle size modern browser that supports many of the latest features in JavaScript. This is where differential Loading comes into play.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: penalty of increased bundle size *when used in a* modern browser

File: aio/content/guide/deployment.md [hunk 4]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+There is a cost to ensure this browser compatibility, and it comes in the form of bundle size. All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't. To maximize compatibility, you ship a single bundle that includes all your transpiled code, plus any polyfills that may be needed. 
+
+Your web application shouldn't have to keep paying the penalty of increased bundle size modern browser that supports many of the latest features in JavaScript. This is where differential Loading comes into play. 
+
+Differential loading is a strategy where you provide two separate bundles as part of your deployed application. One bundle takes advantage of built-in support in modern browsers, ships less polyfills, resulting in a smaller bundle size. The second bundle, includes the additional transpiled code, all necessary polyfills. This strategy allows you to continue to build your web application to support multiple browsers, but only serve up the necessary code that the browsers need.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: you provide -> the CLI will build

File: aio/content/guide/deployment.md [hunk 5]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+### Differential builds
+
+The Angular CLI handles differential loading for you as part of the build process. The Angular CLI will produce the necessary bundles used for differential loading, based on your browser support requirements and compilation target. 
+
+The Angular CLI uses two configurations for differential loading.
+
+- Browserslist - The `browserslist` configuration file is included in your application [project structure](guide/file-structure#application-configuration-files) and provides the minimum browsers your application supports.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: add pointer to the browserslist spec https://github.com/browserslist/browserslist

File: aio/content/guide/deployment.md [hunk 6]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+- tsconfig.json - The `target` in the TypeScript `compilerOptions` determines the ECMAScript target version that the code is transpiled to. Modern browsers support ES2015 natively, while ES5 is more commonly used to support legacy browsers.
+
+The CLI queries the Browserslist configuration, and looks at the `target` to determine if support for legacy browsers is required. The combination of these two configurations determines the number of bundles produced when you create a build. When you create a development build and differential loading is enabled, the output produced is simpler and easier to debug, allowing you to rely less on sourcemaps of transpiled code. When you create a production build, the CLI uses the defined configurations above to determine the bundles to build for deployment of your application. The `index.html` is also modified duing a production build to include the script tags that enable differntial loading.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: mention to see the table below, if the prose is hard to follow?

Also the deprecated `es5BrowserSupport` has an interaction with the feature: `false` means never include es5 polyfills at all, `true` means always do es5 polyfills regardless of browser, and unset means let the diff loading logic determine when to include it

File: aio/content/guide/deployment.md [hunk 7]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+### Configuring differential loading
+
+Differential loading for creating builds is already supported and enabled with version 8 and later of the Angular CLI. You can configure how builds are produced based on the mentioned `browserslist` and `tsconfig.json` files in your application project.
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: note: if it's breaking, we might retreat to safety during RC and make it an opt-in feature instead

File: aio/content/guide/deployment.md [hunk 8]
```diff
@@ -369,3 +369,98 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+Look at the default configuration for a newly created Angular application:
+
+<!-- BR: Can we pull these from somewhere? -->
```
Comment (alexeagle) [inline aio/content/guide/deployment.md]: it should be already stamped out in your project as part of ng update (please verify :)

File: aio/content/guide/deployment.md (hunk 7 above)
Comment (brandonroberts) [reply aio/content/guide/deployment.md]: How about just `is already supported`?

File: aio/content/guide/deployment.md (hunk 8 above)
Comment (brandonroberts) [reply aio/content/guide/deployment.md]: Right. I was referring to an example in our docs set.

File: aio/content/guide/deployment.md [hunk 9]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+- tsconfig.json - The `target` in the TypeScript `compilerOptions` determines the ECMAScript target version that the code is compiled to. Modern browsers support ES2015 natively, while ES5 is more commonly used to support legacy browsers.
+
+
+The CLI queries the Browserslist configuration, and checks the `target` to determine if support for legacy browsers is required. The combination of these two configurations determines whether multiple bundles are produced when you create a build. When you create a development build and differential loading is enabled, the output produced is simpler and easier to debug, allowing you to rely less on sourcemaps of compiled code. When you create a production build, the CLI uses the defined configurations above to determine the bundles to build for deployment of your application. The `index.html` is also modified duing a production build to include the script tags that enable differential loading.
```
Comment (manfredsteyer) [inline aio/content/guide/deployment.md]: Actually, differential loading is not used for ng serve. In this case, we just use the target of the tsconfig.json. This leads to less sources maps and compiled code, as mentioned here. I just think, that this description might make ppl think that differential loading also takes happen during ng serve

File: aio/content/guide/deployment.md [hunk 10]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+**Note:** To see which browsers are supported with the above configuration, see which settings meet to your browser support requirements, see the [Browserslist compatibility page](https://browserl.ist/?q=%3E+0.5%25%2C+last+2+versions%2C+Firefox+ESR%2C+not+dead%2C+not+IE+9-11).
+
+</div>
+
+{@a configuration-table }
+
+| ES5 Browserslist Result | ES Target | Build Result |
+| -------- | -------- | -------- |
+| Disabled | es5     | Single build |
```
Comment (manfredsteyer) [inline aio/content/guide/deployment.md]: Perhaps we should mention that `disabled` means in this context that the `browserslist` does not point to ES5 browsers and `enabled` means that it does.

**Review by LayZeeDK — COMMENTED** [review]
How does the CLI-generated code determine at runtime which bundle to load? I can't identify it in the current draft version of docs.

**Comment by trotyl** [comment]
@LayZeeDK It's based on native browser support for ES module, by assuming that a browser supports ES module already supports ES2015.

So that `<script nomodule>` here means `<script not-load-if-support-es2015>`:

https://github.com/angular/angular-cli/blob/c319cd054e4344737e274af41f46ef4d8a1c3aab/tests/legacy-cli/e2e/tests/build/polyfills.ts#L17-L19

File: aio/content/guide/deployment.md [hunk 11]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+    ],
+    "lib": [
+      "es2018",
+      "dom"
+    ]
+  }
+}
+```
+
+By default, legacy browsers such as IE 9-11 are ignored, and the compilation target is ES2015. As a result, this produces two builds, with differential loading enabled. To see the build result for differential loading based on different configurations, refer to the table below.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: If we ignore browsers without ES2015 support, then we'll produce a single build.

Additionally, currently we support Chrome 41 (for SEO reasons), which requires ES5 build.

File: aio/content/guide/deployment.md [hunk 12]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+There is a cost to ensure this browser compatibility, and it comes in the form of larger bundle size. All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't. To maximize compatibility, you ship a single bundle that includes all your compiled code, plus any polyfills that may be needed. 
+
+Your web application shouldn't have to keep paying the penalty of increased bundle size when used in a modern browser that supports many of the latest features in JavaScript. This is where differential Loading comes into play.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: I'd say that users with modern browsers shouldn't pay the price

File: aio/content/guide/deployment.md [hunk 13]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+There is a cost to ensure this browser compatibility, and it comes in the form of larger bundle size. All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't. To maximize compatibility, you ship a single bundle that includes all your compiled code, plus any polyfills that may be needed. 
+
+Your web application shouldn't have to keep paying the penalty of increased bundle size when used in a modern browser that supports many of the latest features in JavaScript. This is where differential Loading comes into play. 
+
+Differential loading is a strategy where the CLI builds two separate bundles as part of your deployed application. One bundle takes advantage of built-in support in modern browsers, ships less polyfills, resulting in a smaller bundle size. The second bundle, includes the additional compiled code, all necessary polyfills. This strategy allows you to continue to build your web application to support multiple browsers, but only serve up the necessary code that the browsers need.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: The modern bundle also contains modern syntax.

File: aio/content/guide/deployment.md [hunk 14]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+### Differential builds
+
+The Angular CLI handles differential loading for you as part of the build process. The Angular CLI will produce the necessary bundles used for differential loading, based on your browser support requirements and compilation target. 
+
+The Angular CLI uses two configurations for differential loading.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: `loading.` -> `loading:` ?

File: aio/content/guide/deployment.md [hunk 15]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+- Browserslist - The `browserslist` configuration file is included in your application [project structure](guide/file-structure#application-configuration-files) and provides the minimum browsers your application supports. See the [Browserslist spec](https://github.com/browserslist/browserslist) for complete configuration options.
+- tsconfig.json - The `target` in the TypeScript `compilerOptions` determines the ECMAScript target version that the code is compiled to. Modern browsers support ES2015 natively, while ES5 is more commonly used to support legacy browsers.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: We should have a comment that we do not recommend users to use target higher than ES2015. See https://github.com/angular/angular-cli/pull/14206.

File: aio/content/guide/deployment.md (hunk 9 above)
Comment (mgechev) [reply aio/content/guide/deployment.md]: Yes, that's a good point. We should clarify that by development build we mean `ng build` and by production `ng build --prod`.

File: aio/content/guide/deployment.md [hunk 16]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+{@a configuration-table }
+
+| ES5 Browserslist Result | ES Target | Build Result |
+| -------- | -------- | -------- |
+| Disabled | es5     | Single build |
+| Enabled  | es5     | Single build w/Conditional Polyfills |
+| Disabled | es2015  | Single build |
+| Enabled  | es2015  | Differential Loading (Two builds w/Conditional Polyfills |
+
+The `es5BrowserSupport` option in your Angular workspace configuration also let's you explicitly set whether polyfills for ES5 support are included in your bundle.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: `es5BrowserSupport` is deprecated

**Review by mgechev — CHANGES_REQUESTED** [review]
We should also mention how to enable differential loading per project - editing the project's `tsconfig.json` and local `browserlist`.

There are also caveats with some old browsers that we should mention. See this [issue](https://github.com/philipwalton/webpack-esnext-boilerplate/issues/1). We can make a summary in a sentence of two and link to the issue.

**Comment by LayZeeDK** [comment]
> @LayZeeDK It's based on native browser support for ES module, by assuming that a browser supports ES module already supports ES2015.
> 
> So that `<script nomodule>` here means `<script not-load-if-support-es2015>`:
> 
> [angular/angular-cli:tests/legacy-cli/e2e/tests/build/polyfills.ts@`c319cd0`#L17-L19](https://github.com/angular/angular-cli/blob/c319cd054e4344737e274af41f46ef4d8a1c3aab/tests/legacy-cli/e2e/tests/build/polyfills.ts#L17-L19)

Thnanks, Trotyl!

I would expect that to be explained in the Angular docs.

File: aio/content/guide/deployment.md [hunk 17]
```diff
@@ -369,3 +369,104 @@ the subfolder is `my/app/` and you should add `<base href="/my/app/">` to the se
...
+
+The `browserslist` looks like this:
+
+```
+> 0.5%
+last 2 versions
+Firefox ESR
+not dead
+not IE 9-11 # For IE 9-11 support, remove 'not'.
+not Chrome 41 # For Googlebot support, remove 'not'.
```
Comment (mgechev) [inline aio/content/guide/deployment.md]: We changed this https://github.com/angular/angular-cli/pull/14212

**Review by mgechev — APPROVED** [review]
